### ISSUE TITLE
DNS timeouts on nodes with multiple pods being placed simultaneously

### ISSUE THREAD

**Issue by chrisghill** [weaveworks/weave#3663]
## What you expected to happen?
DNS and network to be stable regardless of pod placement/churn

## What happened?
First I want to say this issue seems to be unrelated to the conntrack race condition discussed at length here: https://github.com/kubernetes/kubernetes/issues/56903#

We have been seeing intermittent timeouts in our kubernetes cluster for several months. Originally I found the conntrack race condition issue and I've implemented Quentin Machu's weave-tc patch (https://github.com/Quentin-M/weave-tc) which significantly reduced DNS timeouts, but didn't completely eliminate them. Upon further investigation it seems the issue presents itself on nodes where multiple pods are placed simultaneously - specifically CronJobs with a large amount of `parallelism` all being placed on the same node.

For instance, I've created a single-node InstanceGroup called "quarantine" with a taint on it.  Then I create this simple CronJob:

```
apiVersion: batch/v1beta1
kind: CronJob
metadata:
  namespace: default
  name: dns-test-quarantine
spec:
  schedule: "* * * * *"
  jobTemplate:
    spec:
      parallelism: 20
      template:
        metadata:
          name: dns-test-quarantine
        spec:
          affinity:
            nodeAffinity:
              requiredDuringSchedulingIgnoredDuringExecution:
                nodeSelectorTerms:
                - matchExpressions:
                  - key: InstanceGroup
                    operator: In
                    values:
                    - quarantine
          containers:
          - name: dns-test-quarantine
            image: tutum/dnsutils:latest
            command: ["/bin/sh", "-c"]
            args: ["dig www.github.com"]
          restartPolicy: Never
          tolerations:
          - effect: NoSchedule
            key: InstanceGroup
            operator: Equal
            value: quarantine
```
This will place 20 of those pods on the single quarantine node every minute.  Each minute 2-10 of them will timeout.  While I don't know the specific cause, the reason I'm posting this in the weave project is because I've built separate test clusters (via Kops) using flannel and canal and I've not observed this behavior in either of them.

Some additional information:

- I've run CronJob's like the one above with 3 separate DNS lookups (`dig github.com; dig <our internal redshift fqdn>; dig kube-dns.kube-system.svc`) and only the first lookup ever fails.
- I've also run a deployment that runs `dig github.com; sleep 1` in an infinite loop.  It never times out.  Once a pod is healthy and working it seems to stay working.
- Instead of using parallelism I tried running 7 separate single CronJobs but put a `sleep 5` through `sleep 11` before the `dig` on each of them, effectively spacing them out.  This reduced the number of timeouts, but didn't eliminate them.  You can see it in this Kibana screenshot where I searched for the phrase "connection timed out" (the first half had the staggered sleeps, the 2nd half I removed sleeps):
![image](https://user-images.githubusercontent.com/15616541/61503441-0922af00-a995-11e9-936f-9fb9e405aec2.png)
- To further reinforce that this is unrelated to the conntrack issue, you can even see from the screenshot that I tried using `options single-request`, `options single-request-reopen` and `options use-vc` in the container's `/etc/resolv.conf` (its an Ubuntu-based image) but didn't fix the timeouts.

## How to reproduce it?
Build a basic Kubernetes cluster in Kops using `networking: weave`. Here is our cluster config with sensitive data removed:

<details>

```
apiVersion: kops/v1alpha2
kind: Cluster
metadata:
  name: kooper.k8s.local
spec:
  api:
    loadBalancer:
      type: Public
  authorization:
    rbac: {}
  channel: stable
  cloudLabels:
    cluster: kooper.k8s.local
  cloudProvider: aws
  configBase: <redacted>
  encryptionConfig: true
  etcdClusters:
  - etcdMembers:
    - instanceGroup: master-us-west-2a
      name: a
    - instanceGroup: master-us-west-2b
      name: b
    - instanceGroup: master-us-west-2c
      name: c
    name: main
  - etcdMembers:
    - instanceGroup: master-us-west-2a
      name: a
    - instanceGroup: master-us-west-2b
      name: b
    - instanceGroup: master-us-west-2c
      name: c
    name: events
  iam:
    allowContainerRegistry: true
    legacy: false
  kubeAPIServer:
    admissionControl:
    - Initializers
    - NamespaceLifecycle
    - LimitRanger
    - ServiceAccount
    - PersistentVolumeLabel
    - PersistentVolumeClaimResize
    - DefaultStorageClass
    - DefaultTolerationSeconds
    - MutatingAdmissionWebhook
    - ValidatingAdmissionWebhook
    - NodeRestriction
    - ResourceQuota
    - AlwaysPullImages
    - PodSecurityPolicy
    - DenyEscalatingExec
    auditLogMaxAge: 30
    auditLogMaxBackups: 10
    auditLogMaxSize: 100
    auditLogPath: /var/log/kube-apiserver-audit.log
    auditPolicyFile: /srv/kubernetes/audit-policy.yaml
    authenticationTokenWebhookConfigFile: /srv/kubernetes/aws-iam-authenticator/kubeconfig.yaml
    featureGates:
      ExpandPersistentVolumes: "true"
      TTLAfterFinished: "true"
    runtimeConfig:
      admissionregistration.k8s.io/v1alpha1: "true"
      api/all: "true"
  kubeControllerManager:
    featureGates:
      ExpandPersistentVolumes: "true"
      TTLAfterFinished: "true"
    horizontalPodAutoscalerUseRestClients: true
  kubeDNS:
    provider: CoreDNS
  kubelet:
    featureGates:
      ExpandPersistentVolumes: "true"
      ExperimentalCriticalPodAnnotation: "true"
      TTLAfterFinished: "true"
  kubernetesApiAccess:
  - 0.0.0.0/0
  kubernetesVersion: 1.12.9
  masterInternalName: api.internal.kooper.k8s.local
  masterPublicName: api.kooper.k8s.local
  networkCIDR: 172.20.0.0/16
  networking:
    weave:
      mtu: 8912
  nonMasqueradeCIDR: 100.64.0.0/10
  sshAccess:
  - 0.0.0.0/0
  subnets:
  - cidr: 172.20.32.0/19
    name: us-west-2a
    type: Private
    zone: us-west-2a
  - cidr: 172.20.64.0/19
    name: us-west-2b
    type: Private
    zone: us-west-2b
  - cidr: 172.20.96.0/19
    name: us-west-2c
    type: Private
    zone: us-west-2c
  - cidr: 172.20.0.0/22
    name: utility-us-west-2a
    type: Utility
    zone: us-west-2a
  - cidr: 172.20.4.0/22
    name: utility-us-west-2b
    type: Utility
    zone: us-west-2b
  - cidr: 172.20.8.0/22
    name: utility-us-west-2c
    type: Utility
    zone: us-west-2c
  topology:
    dns:
      type: Public
    masters: private
    nodes: private
  fileAssets: 
  - name: audit-policy.yaml
  <redacted - irrelevant>

---

apiVersion: kops/v1alpha2
kind: InstanceGroup
metadata:
  labels:
    kops.k8s.io/cluster: kooper.k8s.local
  name: bastions
spec:
  image: ami-01966754b218e77ee
  machineType: t2.micro
  maxSize: 1
  minSize: 1
  nodeLabels:
    kops.k8s.io/instancegroup: bastions
    InstanceGroup: bastions
  role: Bastion
  subnets:
  - us-west-2a

---

apiVersion: kops/v1alpha2
kind: InstanceGroup
metadata:
  labels:
    kops.k8s.io/cluster: kooper.k8s.local
  name: master-us-west-2a
spec:
  image: ami-01966754b218e77ee
  machineType: t2.large
  maxSize: 1
  minSize: 1
  nodeLabels:
    kops.k8s.io/instancegroup: master-us-west-2a
    InstanceGroup: master-us-west-2a
  role: Master
  subnets:
  - us-west-2a

---

apiVersion: kops/v1alpha2
kind: InstanceGroup
metadata:
  labels:
    kops.k8s.io/cluster: kooper.k8s.local
  name: master-us-west-2b
spec:
  image: ami-01966754b218e77ee
  machineType: t2.large
  maxSize: 1
  minSize: 1
  nodeLabels:
    kops.k8s.io/instancegroup: master-us-west-2b
    InstanceGroup: master-us-west-2b
  role: Master
  subnets:
  - us-west-2b

---

apiVersion: kops/v1alpha2
kind: InstanceGroup
metadata:
  labels:
    kops.k8s.io/cluster: kooper.k8s.local
  name: master-us-west-2c
spec:
  image: ami-01966754b218e77ee
  machineType: t2.large
  maxSize: 1
  minSize: 1
  nodeLabels:
    kops.k8s.io/instancegroup: master-us-west-2c
    InstanceGroup: master-us-west-2c
  role: Master
  subnets:
  - us-west-2c

---

apiVersion: kops/v1alpha2
kind: InstanceGroup
metadata:
  labels:
    kops.k8s.io/cluster: kooper.k8s.local
  name: compute
spec:
  image: ami-01966754b218e77ee
  machineType: c5.2xlarge
  maxSize: 20
  minSize: 3
  rootVolumeSize: 500
  rootVolumeType: gp2
  nodeLabels:
    kops.k8s.io/instancegroup: compute
    InstanceGroup: compute
    cluster-autoscaler/kooper: "true"
  role: Node
  subnets:
  - us-west-2a
  - us-west-2b
  - us-west-2c

---

apiVersion: kops/v1alpha2
kind: InstanceGroup
metadata:
  labels:
    kops.k8s.io/cluster: kooper.k8s.local
  name: quarantine
spec:
  image: ami-01966754b218e77ee
  machineType: c5.2xlarge
  maxSize: 1
  minSize: 1
  rootVolumeSize: 500
  rootVolumeType: gp2
  nodeLabels:
    kops.k8s.io/instancegroup: quarantine
    InstanceGroup: quarantine
  role: Node
  subnets:
  - us-west-2a
  - us-west-2b
  - us-west-2c
  taints:
  - InstanceGroup=quarantine:NoSchedule
```

</details>

Then launch the dns-test CronJob I specified earlier and monitor the logs. You'll see a handful of them fail with timeouts every minute.

## Anything else we need to know?
You can get most of what you need from the cluster definition
Kubernetes
Built using Kops
Running in AWS

## Versions:

```
$ weave version
weave-net: weaveworks/weave-kube:2.5.1
weave-npc: weaveworks/weave-npc:2.5.1
$ docker version
Client:
 Version:           18.06.3-ce
 API version:       1.38
 Go version:        go1.10.3
 Git commit:        d7080c1
 Built:             Wed Feb 20 02:28:26 2019
 OS/Arch:           linux/amd64
 Experimental:      false

Server:
 Engine:
  Version:          18.06.3-ce
  API version:      1.38 (minimum version 1.12)
  Go version:       go1.10.3
  Git commit:       d7080c1
  Built:            Wed Feb 20 02:26:51 2019
  OS/Arch:          linux/amd64
  Experimental:     false
$ uname -a
Linux ip-172-20-74-32 4.9.0-9-amd64 #1 SMP Debian 4.9.168-1 (2019-04-12) x86_64 GNU/Linux
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"12", GitVersion:"v1.12.9", GitCommit:"e09f5c40b55c91f681a46ee17f9bc447eeacee57", GitTreeState:"clean", BuildDate:"2019-05-27T16:08:57Z", GoVersion:"go1.10.8", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"12", GitVersion:"v1.12.9", GitCommit:"e09f5c40b55c91f681a46ee17f9bc447eeacee57", GitTreeState:"clean", BuildDate:"2019-05-27T15:58:45Z", GoVersion:"go1.10.8", Compiler:"gc", Platform:"linux/amd64"}
```

## Logs:
```
$ docker logs weave
```
or, if using Kubernetes:
```
$ kubectl logs -n kube-system <weave-net-pod> weave
INFO: 2019/07/19 02:26:11.163558 Discovered remote MAC 7a:e6:35:ee:86:fe at 2e:50:c9:2c:92:33(ip-172-20-97-223.us-west-2.compute.internal)
INFO: 2019/07/19 02:26:12.212111 Discovered remote MAC 62:4a:0b:18:69:7e at 2e:50:c9:2c:92:33(ip-172-20-97-223.us-west-2.compute.internal)
INFO: 2019/07/19 02:26:13.599820 Discovered remote MAC e2:9b:95:56:30:c4 at 2e:50:c9:2c:92:33(ip-172-20-97-223.us-west-2.compute.internal)
INFO: 2019/07/19 02:26:14.017246 Discovered remote MAC ba:4d:40:bd:37:50 at b2:25:3e:01:ee:ab(ip-172-20-103-249.us-west-2.compute.internal)
INFO: 2019/07/19 02:26:14.120615 Discovered remote MAC 96:dc:f5:d2:1a:28 at 2e:50:c9:2c:92:33(ip-172-20-97-223.us-west-2.compute.internal)
INFO: 2019/07/19 02:26:14.315328 Discovered remote MAC fe:61:22:49:fa:b0 at b2:25:3e:01:ee:ab(ip-172-20-103-249.us-west-2.compute.internal)
INFO: 2019/07/19 02:26:14.475600 Discovered remote MAC 26:37:42:b1:01:97 at b2:25:3e:01:ee:ab(ip-172-20-103-249.us-west-2.compute.internal)
INFO: 2019/07/19 02:26:14.715802 Discovered remote MAC 26:70:6b:50:dd:25 at b2:25:3e:01:ee:ab(ip-172-20-103-249.us-west-2.compute.internal)
INFO: 2019/07/19 02:26:15.182651 Discovered remote MAC 2e:f7:69:e4:90:ae at b2:25:3e:01:ee:ab(ip-172-20-103-249.us-west-2.compute.internal)
INFO: 2019/07/19 02:26:15.383849 Discovered remote MAC ea:43:2f:ee:3c:29 at 2e:50:c9:2c:92:33(ip-172-20-97-223.us-west-2.compute.internal)
INFO: 2019/07/19 02:26:15.763546 Discovered remote MAC 1a:a2:41:a9:d7:38 at b2:25:3e:01:ee:ab(ip-172-20-103-249.us-west-2.compute.internal)
INFO: 2019/07/19 02:26:16.083735 Discovered remote MAC 4a:2f:46:15:a8:a2 at b2:25:3e:01:ee:ab(ip-172-20-103-249.us-west-2.compute.internal)
INFO: 2019/07/19 02:26:16.456225 Discovered remote MAC 8a:c3:f4:f4:2b:4a at 2e:50:c9:2c:92:33(ip-172-20-97-223.us-west-2.compute.internal)
INFO: 2019/07/19 02:26:17.951117 Discovered remote MAC 3a:61:62:c7:4a:fe at 2e:50:c9:2c:92:33(ip-172-20-97-223.us-west-2.compute.internal)
WARN: 2019/07/19 02:26:26.472216 [allocator]: Delete: no addresses for 5c1532f0ef4f0ab8844fc679c4b08eda350c7de9617a34665d29980cd17bb043
WARN: 2019/07/19 02:26:27.441683 [allocator]: Delete: no addresses for 18c23a8823ba88c9b4c041cfcf695f0481407a22e37c5dac205260d9858d622e
WARN: 2019/07/19 02:26:28.234217 [allocator]: Delete: no addresses for d2d9223b9c507f8f650cc08de146fc35a76fd7104575f008e92767aa450c833a
WARN: 2019/07/19 02:26:31.380017 [allocator]: Delete: no addresses for eb793cc64b2d84a6e7aaa1a59388211457aaf008dcb447678c290cf696c3c798
WARN: 2019/07/19 02:26:34.079396 [allocator]: Delete: no addresses for a6b552a55b5d02069f8e795ec504fe47e91d920abe0216d79052fc6c4d3bba26
WARN: 2019/07/19 02:26:35.978775 [allocator]: Delete: no addresses for 2561103d319a20d4823038f7ac953635483e29975105dfd34743518fb7dc51f9
```
That was during a minute with a bunch of timeouts.
$ journalctl -u docker.service --no-pager
$ journalctl -u kubelet --no-pager
$ kubectl get events
-->

## Network:

```
$ ip route
default via 172.20.96.1 dev eth0 
100.96.0.0/11 dev weave proto kernel scope link src 100.122.128.0 
172.17.0.0/16 dev docker0 proto kernel scope link src 172.17.0.1 linkdown 
172.20.96.0/19 dev eth0 proto kernel scope link src 172.20.103.249
$ ip -4 -o addr
1: lo    inet 127.0.0.1/8 scope host lo\       valid_lft forever preferred_lft forever
2: eth0    inet 172.20.103.249/19 brd 172.20.127.255 scope global eth0\       valid_lft forever preferred_lft forever
3: docker0    inet 172.17.0.1/16 brd 172.17.255.255 scope global docker0\       valid_lft forever preferred_lft forever
6: weave    inet 100.122.128.0/11 brd 100.127.255.255 scope global weave\       valid_lft forever preferred_lft forever
$ sudo iptables-save
```

I put in logs/command that were easy to fetch or didn't have sensitive data.  If there is other data you need that would be relevant I'm happy to get it.

**Comment by murali-reddy** [comment]
@chrisghill sounds similar to https://github.com/weaveworks/weave/issues/3464

Once a pod is started for security reasons traffic is dropped by default till network policies are applied and traffic is allowed explicity. Can you please check weave-npc container logs and verify if there are any dropped packets due to weave npc?

**Comment by chrisghill** [comment]
@murali-reddy Yes! I'm not sure how I didn't check this or notice it before, but I'm seeing stuff like this:

```
WARN: 2019/07/19 04:39:25.928494 UDP connection from 100.110.128.7:45854 to 100.125.128.4:53 blocked by Weave NPC.
WARN: 2019/07/19 04:39:25.928526 UDP connection from 100.110.128.9:37450 to 100.125.128.4:53 blocked by Weave NPC.
WARN: 2019/07/19 04:39:25.928536 UDP connection from 100.110.128.18:44353 to 100.125.0.7:53 blocked by Weave NPC.
WARN: 2019/07/19 04:39:25.928545 UDP connection from 100.110.128.4:39238 to 100.118.0.3:53 blocked by Weave NPC.
WARN: 2019/07/19 04:39:25.928572 UDP connection from 100.110.128.11:35910 to 100.125.128.4:53 blocked by Weave NPC.
WARN: 2019/07/19 04:39:25.928582 UDP connection from 100.110.128.12:35282 to 100.125.0.7:53 blocked by Weave NPC.
```

In the weave-npc logs. I actually saw that exact issue a while back, but I assumed it was unrelated since we don't have any network policies in our cluster (that I'm aware of at least):

```
$ kubectl get networkpolicy --all-namespaces
No resources found.
```

It sounds like this might be the cuplrit, but I don't see any mitigation strategies discussed in the other issue. Do you have any suggestions?

**Comment by murali-reddy** [comment]
Its sort of fixed in https://github.com/weaveworks/weave/pull/3639 targetted for 2.6, so only if you have network policies are applied then you will see this behaviour.

For now you may try adding delay or disable weave-npc if you dont intend to use. Long term clean fix would be use something like podready++ as mentioned in https://github.com/weaveworks/weave/issues/3464#issuecomment-443778814

**Comment by murali-reddy** [comment]
closing this as duplicate of #3464
